src/util/CryptoAmount.ts

```diff
@@ -0,0 +1,148 @@
+import { div, mul } from 'biggystring'
+import { EdgeCurrencyConfig, EdgeDenomination } from 'edge-core-js'
+
+import { RootState } from '../reducers/RootReducer'
+import { emptyEdgeDenomination } from '../selectors/DenominationSelectors'
+import { convertCurrency } from '../selectors/WalletSelectors'
+import { asBiggystring } from './cleaners'
+import { DECIMAL_PRECISION, mulToPrecision } from './utils'
+
+// Defines base properties for asset configuration.
+interface AssetBaseArgs {
+  currencyConfig: EdgeCurrencyConfig
+}
+
+// Specifies that one of tokenId or currencyCode must be provided, or neither.
+type TokenOrCurrencyCodeArgs =
+  | {
+      tokenId: string
+      currencyCode?: never
+    }
+  | {
+      tokenId?: never
+      currencyCode: string
+    }
+  | {
+      tokenId?: never
+      currencyCode?: never
+    }
```

Comment:
During the 4.0 sprint, Paul did a huge amount of work to make `tokenId: string | null` everywhere, so folks *must* to provide it, even if it's just `null` for the parent currency. We had a number of bugs where people simply forgot to pass it because it was optional, and we got the wrong coin. With that in mind, this should be:

```ts

type TokenOrCurrencyCodeArgs =
  | {
      tokenId: EdgeTokenId
      currencyCode?: never
    }
  | {
      tokenId?: never
      currencyCode: string
    }
```

That will force the user to provide one or the other, but not both.

Reply:
I actually started with what you suggested but was opting to appease current callers. I can see how this impl would be dangerous in the same way as the issue you mentioned in the future.